src/main/java/com/github/cbl/algorithm_analyzer/graphs/deepsearch/Deepsearch.java

```diff
@@ -0,0 +1,131 @@
+            implements Event {
+        @Override
+        public String toString() {
+            Object[][] table = new Object[3][];
+            table[0] = new Object[discovered.length + 1];
+            table[1] = new Object[discovered.length + 1];
+            table[2] = new Object[discovered.length + 1];
+            table[0][0] = "";
+            table[1][0] = "Entdeckt";
+            table[2][0] = "Fertig";
+
+            for (int i = 0; i < discovered.length; i++) {
+                table[0][i + 1] = data.nodeNames()[i];
+                table[1][i + 1] = discovered[i];
+                table[2][i + 1] = finished[i];
+            }
+
+            return TablePrinter.toString(table);
+        }
+    }
+    ;
+
+    public static record FinalEdgeTypesEvent(Data data, EdgeType[][] types) implements Event {
+        @Override
+        public String toString() {
+            final StringJoiner sj = new StringJoiner("\n");
+
+            for (int u = 0; u < types.length; u++) {
+                for (int j = 0; j < types.length; j++) {
+                    if (data.graph().hasEdge(u, j)) {
+                        sj.add(
+                                "["
+                                        + data.nodeNames()[u]
+                                        + " -> "
+                                        + data.nodeNames()[j]
+                                        + "]: "
+                                        + types[u][j]);
+                    }
+                }
+            }
+
+            return sj.toString();
+        }
+    }
+
+    public void run(EventConsumer<Event> events, Data data) {
+        Graph<Integer, boolean[][]> graph = data.graph();
+        boolean known[] = new boolean[graph.getVerticeCount()];
+        int previous[] = new int[graph.getVerticeCount()];
+        int discovered[] = new int[graph.getVerticeCount()];
+        int finished[] = new int[graph.getVerticeCount()];
+        int time = 0;
+
+        for (int u = 0; u < graph.getVerticeCount(); u++) {
+            if (!known[u]) {
+                time = this.expand(graph, u, discovered, finished, previous, known, time);
+            }
+        }
+
+        events.accept(new FinalTimesEvent(data, discovered, finished));
+
+        EdgeType[][] types = new EdgeType[graph.getVerticeCount()][];
+
+        for (int u = 0; u < graph.getVerticeCount(); u++) {
+            types[u] = new EdgeType[graph.getVerticeCount()];
+            for (int j = 0; j < graph.getVerticeCount(); j++) {
+                types[u][j] = this.getEdgeType(u, j, graph, discovered, finished, previous);
+            }
+        }
+
+        events.accept(new FinalEdgeTypesEvent(data, types));
+    }
+
+    protected EdgeType getEdgeType(
+            int from,
+            int to,
+            Graph<Integer, boolean[][]> graph,
+            int[] discovered,
+            int[] finished,
+            int[] previous) {
+        if (!graph.hasEdge(from, to)) {
```

Comment:
Just a minor note:
When handling multiple mutual exclusive cases, why not wrap it all together in a form of

```java
if (…) {
…
} else if (…) {
…
} else if (…) {
…
} else {
…
}
```

Eases the reading flow a little bit, in my opinion.
But of course not a real problem here, at worst a very slight code smell (but again: talking about opinions, not facts).

Reply:
Personally, I prefer early returns to if-else conditions, although I haven't been consistent here either.

Reply:
I added early returns for all conditions for consistency https://github.com/cbl/algorithm-analyzer/commit/ac6862303d12ff75565f0eed68858cd4e0e67631